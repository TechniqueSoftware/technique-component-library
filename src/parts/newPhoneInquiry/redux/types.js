export const key = 'newPhoneInquiry'

const SUBMIT = {
  SEND: `${key}/SUBMIT/SEND`,
  ERROR: `${key}/SUBMIT/ERROR`,

  RECEIVE_SUCCESS: `${key}/SUBMIT/RECEIVE_SUCCESS`,
  RECEIVE_ERROR: `${key}/SUBMIT/RECEIVE_ERROR`,

  RECEIVE_APPT_SUCCESS: `${key}/SUBMIT/RECEIVE_APPT_SUCCESS`,
  RECEIVE_APPT_ERROR: `${key}/SUBMIT/RECEIVE_APPT_ERROR`
}

const DATA = {
  SET: `${key}/DATA/SET`,
  TOGGLE_APPOINTMENT: `${key}/DATA/TOGGLE_APPOINTMENT`,

  CLEAR: `${key}/DATA/CLEAR`,

  REQUEST_USER_DATA: `${key}/DATA/REQUEST_USER_DATA`,
  RECEIVE_USER_DATA: `${key}/DATA/RECEIVE_USER_DATA`,

  REQUEST_EMPLOYEES_DATA: `${key}/DATA/REQUEST_EMPLOYEES_DATA`,
  RECEIVE_EMPLOYEES_DATA: `${key}/DATA/RECEIVE_EMPLOYEES_DATA`,

  REQUEST_SCRIPT_DATA: `${key}/DATA/REQUEST_SCRIPT_DATA`,
  RECEIVE_SCRIPT_DATA: `${key}/DATA/RECEIVE_SCRIPT_DATA`,

  SELECT_LOCATION: `${key}/DATA/SELECT_LOCATION`
}

export default {
  SUBMIT,
  DATA
}
