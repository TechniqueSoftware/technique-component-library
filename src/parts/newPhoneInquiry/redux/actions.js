import TYPES from './types'

const SUBMIT_ACTIONS = {
  SEND: () => ({ type: TYPES.SUBMIT.SEND }),
  ERROR: payload => ({ type: TYPES.SUBMIT.ERROR, payload }),

  RECEIVE_SUCCESS: payload => ({ type: TYPES.SUBMIT.RECEIVE_SUCCESS, payload }),
  RECEIVE_ERROR: payload => ({ type: TYPES.SUBMIT.RECEIVE_ERROR, payload }),

  RECEIVE_APPT_SUCCESS: payload => ({ type: TYPES.SUBMIT.RECEIVE_APPT_SUCCESS, payload }),
  RECEIVE_APPT_ERROR: payload => ({ type: TYPES.SUBMIT.RECEIVE_APPT_ERROR, payload })
}

const DATA_ACTIONS = {
  SET: payload => ({ type: TYPES.DATA.SET, payload }),
  TOGGLE_APPOINTMENT: payload => ({ type: TYPES.DATA.TOGGLE_APPOINTMENT, payload }),

  CLEAR: () => ({ type: TYPES.DATA.CLEAR }),
  REQUEST_USER_DATA: payload => ({ type: TYPES.DATA.REQUEST_USER_DATA, payload }),
  RECEIVE_USER_DATA: payload => ({ type: TYPES.DATA.RECEIVE_USER_DATA, payload }),

  REQUEST_EMPLOYEES_DATA: payload => ({ type: TYPES.DATA.REQUEST_EMPLOYEES_DATA, payload }),
  RECEIVE_EMPLOYEES_DATA: payload => ({ type: TYPES.DATA.RECEIVE_EMPLOYEES_DATA, payload }),

  REQUEST_SCRIPT_DATA: payload => ({ type: TYPES.DATA.REQUEST_SCRIPT_DATA, payload }),
  RECEIVE_SCRIPT_DATA: payload => ({ type: TYPES.DATA.RECEIVE_SCRIPT_DATA, payload }),

  SELECT_LOCATION: payload => ({ type: TYPES.DATA.SELECT_LOCATION, payload })
}

export default {
  SUBMIT: SUBMIT_ACTIONS,
  DATA: DATA_ACTIONS
}
