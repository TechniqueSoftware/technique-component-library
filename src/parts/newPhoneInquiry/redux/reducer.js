import { produce } from 'immer'

import moment from 'moment-timezone'
import TYPES from './types'

export const initialState = {
  errors: [],
  appointmentEnabled: false,
  form: {
    locationSelect: {
      type: 'select',
      label: 'Location',
      value: {
        name: '',
        value: '',
        timeZone: ''
      },
      error: '',
      required: true,
      relatedField: null,
      validated: false
    },
    firstName: {
      type: 'text',
      label: 'First Name',
      value: '',
      error: '',
      required: true,
      relatedField: null,
      validated: false
    },
    lastName: {
      type: 'text',
      label: 'Last Name',
      value: '',
      error: '',
      required: true,
      relatedField: null,
      validated: false
    },
    email: {
      type: 'email',
      label: 'Email',
      value: '',
      error: '',
      required: false,
      relatedField: 'mobilePhone',
      validated: false
    },
    mobilePhone: {
      type: 'phone',
      label: 'Phone',
      value: '',
      error: '',
      required: false,
      relatedField: 'email',
      validated: false
    },
    contact: {
      type: 'select',
      label: 'With',
      value: {
        name: '',
        value: '',
        availability: []
      },
      error: '',
      relatedField: null,
      validated: false
    },
    date: {
      type: 'date',
      label: 'Date',
      value: moment().format('YYYY-MM-DD'),
      error: '',
      relatedField: null,
      validated: false
    },
    time: {
      type: 'select',
      label: 'Time',
      value: {
        name: '',
        value: ''
      },
      error: '',
      relatedField: null,
      validated: false
    }
  },
  tiScript: {
    value: 'Loading...'
  },
  response: {}
}

const reducer = (state = initialState, { type, payload }) => produce(state, draft => {
  switch (type) {
    case TYPES.SUBMIT.SEND:
      draft.loading = true
      break
    case TYPES.SUBMIT.ERROR:
      draft.response = payload
      break
    case TYPES.SUBMIT.RECEIVE_SUCCESS:
      draft.loading = false
      draft.response = payload
      break
    case TYPES.SUBMIT.RECEIVE_ERROR:
      draft.loading = false
      draft.response = payload
      break

    case TYPES.DATA.SET:
      draft.form[payload.name] = { ...draft.form[payload.name], ...payload.data }
      break
    case TYPES.DATA.TOGGLE_APPOINTMENT:
      draft.appointmentEnabled = payload
      break

    case TYPES.DATA.CLEAR:
      draft.response = {}
      draft.appointmentEnabled = false
      draft.form = {
        ...initialState.form,
        locationSelect: draft.form.locationSelect,
        contact: draft.form.contact,
        date: draft.form.date,
        time: draft.form.time
      }
      break
    case TYPES.DATA.RECEIVE_USER_DATA:
      draft.response.user = payload
      break
    case TYPES.DATA.RECEIVE_EMPLOYEES_DATA:
      draft.locationEmployees = payload
      break
    case TYPES.DATA.RECEIVE_SCRIPT_DATA:
      draft.tiScript.value = payload
      break
  }
})

export default reducer
