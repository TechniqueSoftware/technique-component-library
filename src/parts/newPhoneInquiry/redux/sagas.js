import { call, put, select, takeEvery, takeLatest } from 'redux-saga/effects'
import axios from '@axios'
import moment from 'moment-timezone'

import { GLOBAL_SELECTOR } from '@redux/global'

import SELECTORS from './selectors'
import ACTIONS from './actions'
import TYPES from './types'

const formatAppointmentTime = (time, timeZone) => {
  if (!time.value?.value) return [null, null]
  const utcStart = moment.tz(time.value.value, timeZone).utc().format()
  const date = new Date(utcStart)
  const utcEnd = new Date(date.getTime() + 30 * 60000).toISOString()
  return [utcStart, utcEnd]
}

const formatFormData = formData => Object.keys(formData).reduce((obj, key) => {
  if (!['locationSelect'].includes(key)) return { ...obj, [key]: formData[key].value }
  return obj
}, {})

const simplifyErrorStructure = err => {
  if (err.status === 400) return { status: err.status, code: err.code, user: err.user }
  if (err.status === 'BAD_REQUEST') {
    let message = ''
    if (err.messages.code.includes('DATA TOO LONG')) {
      if (err.messages.code.includes('FIRST_NAME')) message = 'First name cannot exceed 50 characters'
      if (err.messages.code.includes('LAST_NAME')) message = 'Last name cannot exceed 50 characters'
    }
    return { status: 400, code: err.status, error: err.error, message }
  }
}

export const apiCalls = {
  submit: formBody => axios.post('/users', formBody),
  getUserData: userId => axios.get(`/users/${userId}?limited=true`),
  getLocationData: locationId => axios.get(`/locations/?locationId=${locationId}`),
  getScriptData: locationId => axios.get(`/settings/ti-script/${locationId}`),
  getEmployeesAvailability: ({ locationId, params }) => axios.get(`/locations/${locationId}/availability?eventTypeId=0&eventTypeId=4`, { params }), // eslint-disable-line max-len
  getOwnAvailability: ({ userId, params }) => axios.get(`/users/${userId}/availability?eventTypeId=0&eventTypeId=4`, { params }), // eslint-disable-line max-len
  createAppointmentEvent: appointmentBody => axios.post('/events', appointmentBody)
}

export function* watchGetUserData (action) {
  try {
    if (!action?.payload) throw new Error('There is no user')
    const userData = yield call(apiCalls.getUserData, action.payload)
    yield put(ACTIONS.DATA.RECEIVE_USER_DATA(userData))
  } catch (err) {
    console.error(err) // eslint-disable-line no-console
  }
}

export function* formSubmission () {
  try {
    const user = yield select(GLOBAL_SELECTOR.USER)
    const selectedLocation = yield select(SELECTORS.SELECTED_LOCATION)
    const formData = yield select(SELECTORS.FORM_DATA)
    const appointment = yield select(SELECTORS.APPOINTMENT)
    const appointmentEnabled = yield select(SELECTORS.APPOINTMENT_ENABLED)

    const updatedFormData = appointmentEnabled
      ? formData
      : {
        ...formData,
        contact: {
          value: {
            value: user.userId,
            name: `${user.firstName} ${user.lastName}`
          }
        }
      }

    const userBody = {
      ...formatFormData(updatedFormData),
      locationId: selectedLocation?.value?.value,
      salespersonId: appointmentEnabled ? appointment?.contact?.value.value : user.userId,
      clubId: user.clubId,
      gender: 'M',
      role: { id: 99, name: 'Prospect' },
      source: 'T.I.',
      userOrigin: {
        originId: 2,
        originName: 'T.I.'
      },
      sendResponseEmails: false,
      smsOptOut: true,
      emailOptOut: false
    }

    const data = yield call(apiCalls.submit, JSON.stringify(userBody))
    data.status = 202
    data.code = 'SUCCESS'

    if (appointmentEnabled) {
      const { timeZone, value } = selectedLocation.value
      const [startTime, endTime] = formatAppointmentTime(appointment.time, timeZone)
      const appointmentBody = {
        clubId: user.clubId,
        locationId: value,
        eventTypeId: 4,
        eventOwnerId: appointment?.contact?.value.value,
        startTime,
        endTime,
        subject: null,
        createdById: user.userId,
        remindAttendee: true,
        remindOwner: true,
        bookingSource: 'Club OS T.I. Form',
        attendee: [
          {
            attendeeId: data.userId,
            attendeeStatus: 'A',
            visitType: 'A'
          }
        ]
      }

      const appointmentResponse = yield call(apiCalls.createAppointmentEvent, JSON.stringify(appointmentBody))

      yield put(ACTIONS.SUBMIT.RECEIVE_APPT_SUCCESS(appointmentResponse))
      yield put(ACTIONS.DATA.CLEAR())
    }
    yield put(ACTIONS.SUBMIT.RECEIVE_SUCCESS(data))
  } catch (err) {
    const modifiedError = simplifyErrorStructure(err)

    yield put(ACTIONS.SUBMIT.RECEIVE_ERROR(modifiedError))
    yield put(ACTIONS.SUBMIT.RECEIVE_APPT_ERROR(modifiedError))
  }
}

export function* selectLocation ({ payload }) {
  try {
    const { name, error, validated, value } = payload
    const locationId = value.value
    const dataResponse = yield call(apiCalls.getLocationData, locationId)
    const timeZone = dataResponse?.content[0]?.timeZone
    yield put(
      ACTIONS.DATA.SET({
        name,
        data: { value: { ...value, timeZone }, error, validated }
      })
    )
  } catch (err) {
    return null
  }
}

export function* watchScriptData () {
  try {
    const { value } = yield select(SELECTORS.SELECTED_LOCATION)
    const locationId = value.value

    if (locationId) {
      const scriptData = yield call(apiCalls.getScriptData, locationId)
      yield put(ACTIONS.DATA.RECEIVE_SCRIPT_DATA(scriptData?.script))
    }
  } catch (err) {
    console.error(err) // eslint-disable-line no-console
  }
}

export function* getEmployees () {
  try {
    const selectedLocation = yield select(SELECTORS.SELECTED_LOCATION)
    const { date } = yield select(SELECTORS.APPOINTMENT)
    const { userId } = yield select(GLOBAL_SELECTOR.USER)
    const { value: { value, timeZone } } = selectedLocation
    const locationId = value
    const offset = moment.tz(timeZone).format('Z')
    const params = {
      availabilityDuration: 30,
      startTimeStart: `${date.value}T00:00:00.${offset}`,
      startTimeEnd: `${date.value}T23:45:00.${offset}`,
      multiBookingEventTypeId: 4
    }
    const employeesParams = {
      ...params,
      status: 'available'
    }
    const ownParams = {
      ...params,
      locationId,
      applyBookingThreshold: false
    }
    const employeesRes = yield call(apiCalls.getEmployeesAvailability, { locationId, params: employeesParams })
    const ownAvailabilityRes = yield call(apiCalls.getOwnAvailability, { userId, params: ownParams })
    const users = employeesRes.users.reduce((acc, user) => {
      if (user.userId === userId) return acc.concat(ownAvailabilityRes)
      if (!user.availability.length) return acc
      return acc.concat(user)
    }, [])

    const sortedUsers = [...users].sort((a, b) => a.firstName.localeCompare(b.firstName))
    yield put(ACTIONS.DATA.RECEIVE_EMPLOYEES_DATA(sortedUsers))
  } catch (err) {
    console.error(err) // eslint-disable-line no-console
  }
}

export default function* saga () {
  yield takeEvery(TYPES.DATA.REQUEST_USER_DATA, watchGetUserData)
  yield takeLatest(TYPES.SUBMIT.SEND, formSubmission)
  yield takeLatest(TYPES.DATA.SELECT_LOCATION, selectLocation)
  yield takeLatest(TYPES.DATA.REQUEST_SCRIPT_DATA, watchScriptData)
  yield takeLatest(TYPES.DATA.REQUEST_EMPLOYEES_DATA, getEmployees)
}
