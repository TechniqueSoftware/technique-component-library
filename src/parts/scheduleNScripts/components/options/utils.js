export const followUpTimeToObject = followUpTime => {
  const getHour = h => {
    if (h === '00') return 12
    const parsedHour = parseInt(h, 10)
    return parsedHour > 12 ? parsedHour - 12 : parsedHour
  }
  const time = followUpTime.split(':')
  const [oldHour, oldMinute] = time
  const hour = getHour(oldHour)
  const minute = parseInt(oldMinute, 10)
  const meridiem = oldHour >= 12 ? 'PM' : 'AM'
  return { hour, minute, meridiem }
}

export const objectToFollowUpTime = (hour, minute, meridiem) => {
  const parseHour = h => meridiem === 'PM' ? parseInt(h, 10) + 12 : h
  return `${parseHour(hour === 12 ? '00' : hour)}:${minute === 0 ? '00' : minute}:00`
}
