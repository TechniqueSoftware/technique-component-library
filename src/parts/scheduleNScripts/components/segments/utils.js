import { FOLLOWUP_STATUS } from './constants'

export const parseTitle = followUpStatus => {
  switch (followUpStatus) {
    case FOLLOWUP_STATUS.WEB_LEAD:
      return 'Web Lead'
    case FOLLOWUP_STATUS.APPT_BOOKED:
      return 'Appt Booked'
    case FOLLOWUP_STATUS.MEMBER:
      return 'Member'
  }
}

export const parseLocations = segmentLocations => segmentLocations.map(({ locationName, locationId }) => ({
  label: `${locationName}`,
  value: locationId
}))

export const checkPermission = userPermission => userPermission.some(
  element => element.permissionName === 'CopySegmentsAction' && element.permissionEffect === 'ALLOW')
