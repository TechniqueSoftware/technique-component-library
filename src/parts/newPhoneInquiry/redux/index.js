import { createManager } from '@redux/injector'

import reducer from './reducer'
import { key } from './types'
import saga from './sagas'

export { default as REDUCER } from './reducer'
export { default as ACTIONS } from './actions'
export { default as SELECTORS } from './selectors'
export { key, default as TYPES } from './types'

export const Manager = createManager({
  name: 'NewPhoneInquiryManager',
  key,
  reducer,
  saga
})
