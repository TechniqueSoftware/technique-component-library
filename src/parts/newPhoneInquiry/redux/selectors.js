import { createSelector } from 'reselect'
import { get } from 'lodash'

import { key } from './types'
import { initialState } from './reducer'

const reducer = state => get(state, key, initialState)

const getFormResponse = createSelector(
  reducer, state => get(state, 'response', {})
)

const getForm = createSelector(
  reducer, state => get(state, 'form', initialState.form)
)

const getConstants = createSelector(
  reducer, state => get(state, 'constants', initialState.formConstants)
)

const getScriptData = createSelector(
  reducer, state => get(state, 'tiScript[value]', '')
)

const getEmployees = createSelector(
  reducer, state => get(state, 'locationEmployees', [])
)

const getAppointment = createSelector(reducer, state => {
  const form = get(state, 'form', initialState.form)
  return { contact: form.contact, date: form.date, time: form.time }
})

const getAppointmentEnabled = createSelector(
  reducer, state => get(state, 'appointmentEnabled', false)
)

const getSelectedLocation = createSelector(
  reducer, state => get(state, 'form[locationSelect]', initialState.form.locationSelect)
)

export default {
  FORM_RESPONSE: getFormResponse,
  FORM_DATA: getForm,
  FORM_CONSTANTS: getConstants,
  SCRIPT: getScriptData,
  EMPLOYEES: getEmployees,
  APPOINTMENT: getAppointment,
  APPOINTMENT_ENABLED: getAppointmentEnabled,
  SELECTED_LOCATION: getSelectedLocation
}
