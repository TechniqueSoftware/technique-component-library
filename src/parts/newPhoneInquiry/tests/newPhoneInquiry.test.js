import React from 'react'
import * as redux from 'react-redux'
import { render, screen } from '@testing-library/react'

import { TestProvider as Provider } from '@redux/utils'
import userEvent from '@testing-library/user-event'
import Inquiry from '..'

import { state } from './redux/mock'
import { ACTIONS } from '../redux'

jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb(state))
const mockDispatch = jest.fn()
jest.spyOn(redux, 'useDispatch').mockReturnValue(mockDispatch)

describe('<Inquiry/>', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<Provider><Inquiry /></Provider>)
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render the modal button', () => {
    render(<Provider><Inquiry /></Provider>)
    const TIButton = screen.getByText('T.I. Inquiry')
    expect(TIButton).toBeInTheDocument()
  })

  it('opens modal', () => {
    render(<Provider><Inquiry /></Provider>)
    const TIButton = screen.getByText('T.I. Inquiry')
    expect(TIButton).toBeInTheDocument()
    expect(screen.queryByText('Add Telephone Inquiry')).not.toBeInTheDocument()
    userEvent.click(TIButton)
    expect(screen.queryByText('Add Telephone Inquiry')).toBeInTheDocument()
    expect(screen.queryByText('Location')).toBeInTheDocument()
    expect(screen.queryByText('First Name *')).toBeInTheDocument()
    expect(screen.queryByText('Last Name *')).toBeInTheDocument()
  })

  it('saves TI Entry on save button click', async () => {
    render(<Provider><Inquiry /></Provider>)
    const TIButton = screen.getByText('T.I. Inquiry')
    userEvent.click(TIButton)
    userEvent.type(screen.getByLabelText('First Name *'), 'firstName')
    userEvent.type(screen.getByLabelText('Last Name *'), 'lastName')
    userEvent.type(screen.getByLabelText('Email'), 'email@email.com')
    userEvent.click(screen.getByText('Save'))
    expect(mockDispatch).toHaveBeenCalledWith(ACTIONS.SUBMIT.SEND())
  })
})
