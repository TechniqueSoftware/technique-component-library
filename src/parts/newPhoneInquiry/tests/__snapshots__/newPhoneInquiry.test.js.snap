// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Inquiry/> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="inqButton"
  >
    <button
      class="
    d-flex
    align-items-center
    justify-content-center
    flex-row
    button
    
    undefined
   btn btn-primary btn-md"
      role="button"
      type="button"
    >
      <span>
        T.I. Inquiry
      </span>
    </button>
  </div>
</DocumentFragment>
`;
